File: LeetCode/39.Combination_Sum/Mario/39.Combination_Sum.java
class Solution {
    private List<List<Integer>> res;

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        res = new ArrayList<>();
        backtracking(candidates, target, 0, new ArrayList<Integer>(), 0);
        return res;
    }

    private void backtracking(int[] cads, int target, int sum, List<Integer> list, int start) {
        if (target == sum) {
            res.add(new ArrayList<>(list));
            return;
        }

        for (int i = start; i < cads.length; i++) {
            if (sum + cads[i] > target) {
                continue;
            }
            list.add(cads[i]);
            backtracking(cads, target, sum + cads[i], list, i);
            list.remove(list.size() - 1);
        }
    }
}

File: LeetCode/39.Combination_Sum/Mario/39.Combination_Sum.py
#!/usr/bin/env python3


# backtrack
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        n = len(candidates)
        res = []
        path = []

        def backtrack(i, total):
            if total > target:
                return
            if total == target:
                res.append(path[:])
                return

            for j in range(i, n):
                can = candidates[j]
                path.append(can)
                backtrack(j, total + can)
                path.pop()

        backtrack(0, 0)

        return res


# dp
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        # dp[i] = []
        # dp[i] = dp[w - cand[i]] + [cand]

        # init dp
        dp = [[] for _ in range(target + 1)]
        dp[0] = [[]]

        for c in candidates:
            for j in range(c, target + 1):
                for e in dp[j - c]:
                    dp[j].append(e + [c])

        # print(dp)

        return dp[target]
